core/src/main/java/org/elasticsearch/cluster/metadata/IndexCreationMonitor.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import com.carrotsearch.hppc.cursors.IntObjectCursor;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.replication.ReplicationOperation;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateObserver;
+import org.elasticsearch.cluster.ack.ClusterStateUpdateResponse;
+import org.elasticsearch.cluster.routing.IndexRoutingTable;
+import org.elasticsearch.cluster.routing.IndexShardRoutingTable;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+
+import java.util.Objects;
+
+import static org.elasticsearch.action.support.replication.ReplicationOperation.WAIT_ON_ACTIVE_SHARDS_SETTING;
+import static org.elasticsearch.action.support.replication.ReplicationOperation.checkWaitOnActiveShards;
+
+/**
+ * This class is responsible for waiting to send an index creation
+ * response on the action listener until indexing or other replication
+ * actions can successfully be performed on the newly created index.
+ *
+ * It should be subclassed to implement the particular response type
+ * desired for the index creation operation.
+ */
+public final class IndexCreationMonitor extends AbstractComponent {
+
+    private final ClusterService clusterService;
+    private final ThreadContext threadContext;
+
+    protected IndexCreationMonitor(Settings settings, ClusterService clusterService, ThreadContext threadContext) {
```

Comment:
We should use the thread context of the caller of `waitOnShards`, not the one passed here when we construct `IndexCreationMonitor`. Make this `ThreadPool` instead.